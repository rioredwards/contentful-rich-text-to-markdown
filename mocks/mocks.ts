import { Document } from "@contentful/rich-text-types";
import { BLOCKS } from "@contentful/rich-text-types";

export const basicRichText: Document = {
  data: {},
  content: [
    {
      data: {},
      content: [
        {
          data: {},
          marks: [],
          value: "A classic game of Tic Tac Toe for the Command Line.",
          nodeType: "text",
        },
      ],
      nodeType: BLOCKS.PARAGRAPH,
    },
  ],
  nodeType: BLOCKS.DOCUMENT,
};

export const richTextWithList: Document = {
  data: {},
  content: [
    {
      data: {},
      content: [
        {
          data: {},
          content: [
            {
              data: {},
              content: [
                {
                  data: {},
                  marks: [],
                  value: "Designed using OOP principles",
                  nodeType: "text",
                },
              ],
              nodeType: BLOCKS.PARAGRAPH,
            },
          ],
          nodeType: BLOCKS.LIST_ITEM,
        },
        {
          data: {},
          content: [
            {
              data: {},
              content: [
                {
                  data: {},
                  marks: [],
                  value: "Intuitive Command Line Interface",
                  nodeType: "text",
                },
              ],
              nodeType: BLOCKS.PARAGRAPH,
            },
          ],
          nodeType: BLOCKS.LIST_ITEM,
        },
      ],
      nodeType: BLOCKS.UL_LIST,
    },
  ],
  nodeType: BLOCKS.DOCUMENT,
};

export const richTextWithCodeBlock: Document = {
  nodeType: BLOCKS.DOCUMENT,
  data: {},
  content: [
    {
      nodeType: BLOCKS.PARAGRAPH,
      data: {},
      content: [
        { nodeType: "text", value: "1. Download and cd into the project", marks: [], data: {} },
      ],
    },
    {
      nodeType: BLOCKS.EMBEDDED_ENTRY,
      data: {
        target: {
          metadata: { tags: [] },
          sys: {
            space: { sys: { type: "Link", linkType: "Space", id: "l329ngjcm8m3" } },
            id: "6IoowmQ4rIHAP4rHezXIvX",
            type: "Entry",
            createdAt: "2023-10-15T22:36:59.968Z",
            updatedAt: "2023-10-16T04:49:29.650Z",
            environment: { sys: { id: "master", type: "Link", linkType: "Environment" } },
            revision: 5,
            contentType: { sys: { type: "Link", linkType: "ContentType", id: "codeBlock" } },
            locale: "en-US",
          },
          fields: {
            title: "Tic Tac Toe Download",
            content: "git clone https://github.com/rioredwards/Tic-Tac-Toe.git\\ncd Tic-Tac-Toe",
            language: "bash",
          },
        },
      },
      content: [],
    },
    {
      nodeType: BLOCKS.PARAGRAPH,
      data: {},
      content: [
        {
          nodeType: "text",
          value: "2. Compile source code into binary (executable) file",
          marks: [],
          data: {},
        },
      ],
    },
    {
      nodeType: BLOCKS.EMBEDDED_ENTRY,
      data: {
        target: {
          metadata: { tags: [] },
          sys: {
            space: { sys: { type: "Link", linkType: "Space", id: "l329ngjcm8m3" } },
            id: "2WVtiGV7KEvIVNDJoEe50d",
            type: "Entry",
            createdAt: "2023-10-16T04:44:29.573Z",
            updatedAt: "2023-10-16T04:44:29.573Z",
            environment: { sys: { id: "master", type: "Link", linkType: "Environment" } },
            revision: 1,
            contentType: { sys: { type: "Link", linkType: "ContentType", id: "codeBlock" } },
            locale: "en-US",
          },
          fields: { title: "Tic Tac Toe Compile Into Binary", content: "make", language: "bash" },
        },
      },
      content: [],
    },
    {
      nodeType: BLOCKS.PARAGRAPH,
      data: {},
      content: [{ nodeType: "text", value: "3. Run the executable file", marks: [], data: {} }],
    },
    {
      nodeType: BLOCKS.EMBEDDED_ENTRY,
      data: {
        target: {
          metadata: { tags: [] },
          sys: {
            space: { sys: { type: "Link", linkType: "Space", id: "l329ngjcm8m3" } },
            id: "Vv8gPIT5uLr3353UjiGBt",
            type: "Entry",
            createdAt: "2023-10-16T04:45:22.572Z",
            updatedAt: "2023-10-16T04:45:22.572Z",
            environment: { sys: { id: "master", type: "Link", linkType: "Environment" } },
            revision: 1,
            contentType: { sys: { type: "Link", linkType: "ContentType", id: "codeBlock" } },
            locale: "en-US",
          },
          fields: { title: "Tic Tac Toe Run Executable", content: "./main", language: "bash" },
        },
      },
      content: [],
    },
    {
      nodeType: BLOCKS.PARAGRAPH,
      data: {},
      content: [{ nodeType: "text", value: "", marks: [], data: {} }],
    },
  ],
};
