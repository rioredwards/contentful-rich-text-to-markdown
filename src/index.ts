import { documentToHtmlString } from "@contentful/rich-text-html-renderer";
import type { Document } from "@contentful/rich-text-types";
import { NodeHtmlMarkdown, TranslatorConfigObject } from "node-html-markdown";
import { Options } from "@contentful/rich-text-html-renderer";

const formatCodeBlock = (node: any) => {
  const { content, language } = node;
  const markdown = `<CustomCodeBlock>\`\`\`${language}\n${content}\n\`\`\`</CustomCodeBlock>`;
  return markdown;
};

const formatImage = (title: string, url: string) => {
  const markdown = `<CustomEmbeddedImage>![${title}](https:${url})</CustomEmbeddedImage>`;
  return markdown;
};

const htmRendererOptions: Options = {
  renderNode: {
    "embedded-entry-block": (node, _) => {
      if (node?.data) {
        const entry = node.data.target;
        if (entry.sys.contentType.sys.id === "codeBlock") {
          const markdown = formatCodeBlock(entry.fields);
          return markdown;
        }
      }
      return `unhandled node_type: ${node.nodeType}`;
    },
    "embedded-entry-inline": (node, _) => {
      return `unhandled node_type: ${node.nodeType}`;
    },
    "embedded-asset-block": (node, _) => {
      if (node?.data) {
        const entry = node.data.target;
        try {
          const title = entry.fields.title;
          const url = entry.fields.file.url;
          const markdown = formatImage(title, url);
          return markdown;
        } catch (err) {
          console.error("Error rendering embedded-asset-block", err);
        }
      }
      return `unhandled node_type: ${node.nodeType}`;
    },
  },
};

// Set NodeHtmlMarkdown to just print out the innerText of the CustomCodeBlocks
const customTranslatorConfigForCodeBlock: TranslatorConfigObject = {
  CustomCodeBlock: {
    postprocess: (ctx) => ctx.node.innerText,
    preserveWhitespace: true,
    surroundingNewlines: 2,
  },
  CustomEmbeddedImage: {
    postprocess: (ctx) => ctx.node.innerText,
    preserveWhitespace: true,
    surroundingNewlines: 2,
  },
};

// Convert the rich text documents from Contentful to Markdown
export function convertRichTextToMarkdown(document: Document): string | undefined {
  const html = documentToHtmlString(document, htmRendererOptions);
  const markdown = NodeHtmlMarkdown.translate(html, {}, customTranslatorConfigForCodeBlock);
  return markdown;
}
