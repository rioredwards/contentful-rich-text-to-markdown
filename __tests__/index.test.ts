import { convertRichTextToMarkdown } from "../src/index";
import { describe, expect, test } from "@jest/globals";
import { basicRichText, richTextWithList } from "../mocks/mocks";

describe("convertRichTextToMarkdown should return markdown when passed a: ", () => {
  it("basic rich-text obj", () => {
    const markdown = convertRichTextToMarkdown(basicRichText);

    expect(markdown).toEqual("A classic game of Tic Tac Toe for the Command Line.");
  });
  it("rich-text obj with unordered list", () => {
    const markdown = convertRichTextToMarkdown(richTextWithList);

    expect(markdown).toMatchSnapshot();
  });
});
