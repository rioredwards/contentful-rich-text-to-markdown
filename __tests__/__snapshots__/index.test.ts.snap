// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertRichTextToMarkdown should return markdown when passed a:  rich-text obj with code block 1`] = `
"1\\. Download and cd into the project

\`\`\`bash
git clone https://github.com/rioredwards/Tic-Tac-Toe.git\\ncd Tic-Tac-Toe
\`\`\`

2\\. Compile source code into binary (executable) file

\`\`\`bash
make
\`\`\`

3\\. Run the executable file

\`\`\`bash
./main
\`\`\`"
`;

exports[`convertRichTextToMarkdown should return markdown when passed a:  rich-text obj with code embedded asset 1`] = `"{"customEmbeddedImage":{"title":"Test Logo","url":"//images.ctfassets.net/l329ngjcm8m3/61Gv4YS4gh15J9LlabYYGw/c1bb5e48c4fb498e1a83cbcf5ea92085/Test_Logo.png"}}"`;

exports[`convertRichTextToMarkdown should return markdown when passed a:  rich-text obj with unordered list 1`] = `
"* Designed using OOP principles
* Intuitive Command Line Interface"
`;

exports[`convertRichTextToMarkdown should return markdown when passed a:  rich-text obj with unordered list 2`] = `
"* Designed using OOP principles
* Intuitive Command Line Interface"
`;
